File: CSES/Introductory Problems/q24_Grid_Path_Description/toPass2.cpp
#include<bits/stdc++.h>
#define int long long int
#define F first
#define S second
#define pb push_back
#define RIGHT 0
#define LEFT 1
#define DOWN 2
#define UP 3
#define isValid(a,b,c) (a>=b && a < c ? 1: 0)
using namespace std;
int dx[4] = {0,0,1,-1};
int dy[4] = {1,-1,0,0};
string str;
int vis[7][7];
void solve()
{

}
int countPaths(int x,int y, int pos)
{
	if(pos == (int)str.length())
		return (x==6 && y==0);
	if(x == 6 && y == 0)
		return 0;
	if(vis[x][y])
		return 0;
	vector<bool> visited(4,-1);
	for(int k = 0; k < 4; k++)
	{
		if(isValid(x + dx[k],0,7) && isValid(y + dy[k],0,7))
			visited[k] = vis[x + dx[k]][y+dy[k]];
	}
	if(!visited[DOWN] && !visited[UP] && visited[RIGHT] && visited[LEFT])
		return 0;
	if(!visited[RIGHT] && !visited[LEFT] && visited[DOWN] && visited[UP])
		return 0;
	if(isValid(x-1,0,7) && isValid(y+1,0,7) && vis[x-1][y+1]==1)
	{
		if(!visited[RIGHT] && !visited[UP])
		   return 0;
	}
	if(isValid(x+1,0,7) && isValid(y+1,0,7) && vis[x+1][y+1]==1)
	{
		if(!visited[RIGHT] && !visited[DOWN])
		   return 0;
	}
	if(isValid(x-1,0,7) && isValid(y-1,0,7) && vis[x-1][y-1]==1)
	{
		if(!visited[LEFT] && !visited[UP])
		   return 0;
	}
	if(isValid(x+1,0,7) && isValid(y-1,0,7) && vis[x+1][y-1]==1)
	{
		if(!visited[LEFT] && !visited[DOWN])
		   return 0;
	}
	vis[x][y] = 1;
	int numOfPaths = 0; 
	if(str[pos] == '?')
	{
		for(int k = 0; k < 4; k++)
		{
			if(isValid(x+dx[k],0,7) && isValid(y+dy[k],0,7))
				numOfPaths += countPaths(x+dx[k], y+dy[k], pos+1);
		}
	}
	else if(str[pos] == 'R' && y + 1 < 7)
		numOfPaths+= countPaths(x,y+1,pos+1);
	else if(str[pos] == 'L' && y -1 >= 0)
		numOfPaths+= countPaths(x,y-1,pos+1);
	else if(str[pos] == 'U' && x -1 >= 0)
		numOfPaths+= countPaths(x-1,y,pos+1);
	else if(str[pos] == 'D' && x + 1 < 7)
		numOfPaths+= countPaths(x+1,y,pos+1);
	vis[x][y] = 0;
	return numOfPaths;

}
int32_t main()
{
    #ifndef ONLINE_JUDGE
       freopen("input.txt","r",stdin);
       freopen("output.txt","w",stdout);
    #endif
       cin >> str;
       cout << countPaths(0,0,0) << endl;
       
       
}

File: CSES/Introductory Problems/q24_Grid_Path_Description/toPass2.py
dx = [0, 0, 1, -1]  # RIGHT, LEFT, DOWN, UP
dy = [1, -1, 0, 0]

def is_valid(a, b, c):
    return b <= a < c

vis = [[0 for _ in range(7)] for _ in range(7)]

def count_paths(x, y, pos, s):
    if pos == len(s):
        return int(x == 6 and y == 0)
    if x == 6 and y == 0:
        return 0
    if vis[x][y] == 1:
        return 0

    visited = [False] * 4
    for k in range(4):
        nx, ny = x + dx[k], y + dy[k]
        if is_valid(nx, 0, 7) and is_valid(ny, 0, 7):
            visited[k] = vis[nx][ny] == 1

    if not visited[2] and not visited[3] and visited[0] and visited[1]:
        return 0
    if not visited[0] and not visited[1] and visited[2] and visited[3]:
        return 0

    if is_valid(x-1, 0, 7) and is_valid(y+1, 0, 7) and vis[x-1][y+1] == 1:
        if not visited[0] and not visited[3]:
            return 0
    if is_valid(x+1, 0, 7) and is_valid(y+1, 0, 7) and vis[x+1][y+1] == 1:
        if not visited[0] and not visited[2]:
            return 0
    if is_valid(x-1, 0, 7) and is_valid(y-1, 0, 7) and vis[x-1][y-1] == 1:
        if not visited[1] and not visited[3]:
            return 0
    if is_valid(x+1, 0, 7) and is_valid(y-1, 0, 7) and vis[x+1][y-1] == 1:
        if not visited[1] and not visited[2]:
            return 0

    vis[x][y] = 1
    num_paths = 0

    if s[pos] == '?':
        for k in range(4):
            nx, ny = x + dx[k], y + dy[k]
            if is_valid(nx, 0, 7) and is_valid(ny, 0, 7):
                num_paths += count_paths(nx, ny, pos + 1, s)
    else:
        ch = s[pos]
        if ch == 'R' and y + 1 < 7:
            num_paths += count_paths(x, y + 1, pos + 1, s)
        elif ch == 'L' and y - 1 >= 0:
            num_paths += count_paths(x, y - 1, pos + 1, s)
        elif ch == 'U' and x - 1 >= 0:
            num_paths += count_paths(x - 1, y, pos + 1, s)
        elif ch == 'D' and x + 1 < 7:
            num_paths += count_paths(x + 1, y, pos + 1, s)

    vis[x][y] = 0
    return num_paths

if __name__ == "__main__":
    s = input().strip()
    print(count_paths(0, 0, 0, s))
